runtime/lua/lsp/builtin_callbacks.lua

```diff
@@ -0,0 +1,349 @@
+-- luacheck: globals vim
+-- Implements the following default callbacks:
+--  textDocument/publishDiagnostics
+
+--  IN PROGRESS: textDocument/completion
+--  TODO: completionItem/resolve
+
+--  textDocument/hover
+--  TODO: textDocument/signatureHelp
+--  textDocument/references
+--  TODO: textDocument/documentHighlight
+--  TODO: textDocument/documentSymbol
+--  TODO: textDocument/formatting
+--  TODO: textDocument/rangeFormatting
+--  TODO: textDocument/onTypeFormatting
+--  textDocument/definition
+--  TODO: textDocument/codeAction
+--  TODO: textDocument/codeLens
+--  TODO: textDocument/documentLink
+--  TODO: textDocument/rename
+
+--  TODO: codeLens/resolve
+
+--  TODO: documentLink/resolve
+
+local log = require('lsp.log')
+local protocol = require('lsp.protocol')
+local errorCodes = protocol.errorCodes
+
+local QuickFix = require('nvim.quickfix_list')
+local LocationList = require('nvim.location_list')
+local handle_completion = require('lsp.handle.completion')
+local handle_workspace = require('lsp.handle.workspace')
+
+
+-- Callback definition section
+local add_callback = function(callback_mapping, callback_object, name, callback, options)
+  callback_object_instance = callback_object.new(name, options)
+  callback_object_instance:add_callback(callback)
+  callback_mapping[name] = callback_object_instance
+end
+
+-- 3 nvim/error_callback
+local add_nvim_error_callback = function(callback_mapping, callback_object)
+  add_callback(callback_mapping, callback_object, 'nvim/error_callback', function(original, error_message)
+    local message = ''
+    if error_message.message ~= nil and type(error_message.message) == 'string' then
+      message = error_message.message
+    elseif rawget(errorCodes, error_message.code) ~= nil then
+      message = string.format('[%s] %s',
+        error_message.code, errorCodes[error_message.code]
+      )
+    end
+
+    vim.api.nvim_err_writeln(string.format('[LSP:%s] Error: %s', original.method, message))
+
+    return
+  end)
+end
+
+-- 3 textDocument/publishDiagnostics
+local add_text_document_publish_diagnostics_callback = function(callback_mapping, callback_object)
+  add_callback(callback_mapping, callback_object, 'textDocument/publishDiagnostics', function(self, data)
```

Comment:
having an extra layer of "adder" functions seems a bit clumsy. Can't be just have a table of default callbacks and then write
`default_callbacks["textDocument/publishDiagnostics"] = function(self,data) ... end`. Then `add_all_callbacks` can just loop over the table. Or the user can add a subset by `add_callbacks{'textDocument/publishDiagnostics', 'textDocument/completion', ...}` .